### DISCUSSION TITLE
Multiple projects

### DISCUSSION THREAD

**Discussion by MarkoShenodaa** [unoplatform/uno#16781] [Q&A]
Hello,

i want the project to be multiple not one so i cane debug many ones at same time 

![image](https://github.com/unoplatform/uno/assets/137176976/5c65df2c-f1cc-4c95-a878-a447b55f2d19)

is there any way to do that

thank you
marko shenoda

**Comment by agneszitte** [comment]
cc @jeromelaban

**Reply by jeromelaban** [reply]
@MarkoShenodaa thanks for the report. While this is Visual Studio a behavior that we cannot change, it does not prevent from debugging multiple projects at the same time, it only requires that the SLN is not in the same folder as the csproj being created.

If not, you can still move projects once they've been created.

**Reply by agneszitte** [reply]
@jeromelaban do we need or are we missing info in the current documentation regarding these details?

**Reply by jeromelaban** [reply]
> @jeromelaban Thaks for replay.sorry if i'm wasn't clear how i split the plat forms with this new type of project i need to split my project to multiple one for android, one for ios, one for web and one for desktop as example

I see, then you'll need to create separate projects with one target frameworks each. You can use the `dotnet new` command to create each projects with one platform enabled.

Out of curiosity, why do you need to have individual projects?

**Reply by MarkoShenodaa** [reply]
@jeromelaban  i want to build my app using UNO so it save time for me however there is some sitting i want to made for each platform,
i need to test the result in all of them at the same time.

also i need to add some thing for android and other for ios so i need to see the effect at the same time

**Reply by MarkoShenodaa** [reply]
is there any plan to make an option to select one project or multiple 

please note that this is critical for my project plan 

thank you

**Reply by jeromelaban** [reply]
At this time, we do not have plans to create a template that generates multiple projects. You can easily make multiple projects by creating each target one by one.

> i need to test the result in all of them at the same time.

I also do not understand what this means. Can you point to an example that shows what this is, using another public solution or video?

**Reply by MarkoShenodaa** [reply]
i build inventory system with multiple ui configuration based on user login device those setting store on database. when user login some forms have some element based on the device he selected so i need to see the effect on all the devices with on single app i can't run all of them at same time so i can see the effect. by have single project i must stop and run the project for each device.

**Reply by jeromelaban** [reply]
I see. That makes sense. In this case, you can create multiple heads using the templates, and move your app code in a class library.

**Reply by MarkoShenodaa** [reply]
> I see. That makes sense. In this case, you can create multiple heads using the templates, and move your app code in a class library.


so that is mean i create the view for etch platform i can't make it same view?

**Reply by jeromelaban** [reply]
You definitely can, using a class library that contains all your views.

Still, I wonder, when you're saying that you can't see the effect on all devices at the same time, is it because you want to debug on all apps at once? Can't you deploy the app without the debugger on all devices and see the effect on all those apps?

**Reply by MarkoShenodaa** [reply]
> You definitely can, using a class library that contains all your views.
> 
> Still, I wonder, when you're saying that you can't see the effect on all devices at the same time, is it because you want to debug on all apps at once? Can't you deploy the app without the debugger on all devices and see the effect on all those apps?

i want to be able to debug if needed when they run

**Comment by MarkoShenodaa** [comment]
@jeromelaban Thaks for replay.sorry if i'm wasn't clear how i split the plat forms with this new type of project i need to split my project to multiple one for android, one for ios, one for web and one for desktop as example

**Reply by ToddGlodek** [reply]
I, for example, am in a situation where we MUST HAVE A SEPARATE CONTROLS PROJECT because the third party vendor for our critical library does not provide a cross platform library - they have separate native libraries for iOS and Android.  Consequently, we have placed the code related to these libraries in a whole separate repository so that we allow outside developers to access that code without extending them visibility to the entirety of our code base.

It formerly was required in older versions of Uno & Xamarin to maintain heavy head projects for iOS, Android, etc.  - but this is no longer the case.  Because we then use dependency injection to load the appropriate library for each iOS/Android platform, any kind of custom platform-specific library ( i.e. native UI ) we can now loaded and managed using you only need a tiny head project for platform.  Uno will then compile each platform output separately using ONLY the appropriate library for each platform.

One thing that would be very helpful, however, would be if Uno were to change the boiler-plate code that was generated in by the project templates so that it is clear how custom platform specific libraries should be handled.  Currently, Uno templates produce something like this:

```
public class App : Application
{
    protected Window? MainWindow { get; private set; }
    protected IHost? Host { get; private set; }

    protected async override void OnLaunched(LaunchActivatedEventArgs args)
    {
        var builder = this.CreateBuilder(args)
            // Add navigation support for toolkit controls such as TabBar and NavigationView
            .UseToolkitNavigation()
            .Configure(host => host
                .UseLogging(configure: (context, logBuilder) =>
                {
                    // Configure log levels for different categories of logging
                    logBuilder
                        .SetMinimumLevel(
                            context.HostingEnvironment.IsDevelopment() ?
                                LogLevel.Information :
                                LogLevel.Warning)
                 }
                .ConfigureServices((context, services) =>
                {
                    // TODO: Register your services
                    //services.AddSingleton<IMyService, MyService>();
                })                
            );
    }
```
 However, something like this would be much more clear 
 
```
      .ConfigureServices((context, services) =>
      {
          // TODO: Register your services
          //services.AddSingleton<IMyService, MyService>();

          RegisterYourPlatformSpecificServices(context, services);
      })

```
Then add a virtual method that can be overridden in each platform specific head project

```
    protected virtual void RegisterYourPlatformSpecificServices(IHostBuilder hostBuilder, IServiceCollection services)
    {
        // TODO: Register your services
        //services.AddSingleton<IMyService, MyService>();
    }
```
All that would be needed in each project's boiler-plate template would be to override that method in each platforms AppHead class:

```
public sealed partial class AppHead : App
{
    /// <summary>
    /// Initializes the singleton application object. This is the first line of authored code
    /// executed, and as such is the logical equivalent of main() or WinMain().
    /// </summary>
    public AppHead()
    {
        this.InitializeComponent();
    }

    /// <summary>
    /// Invoked when the application is launched normally by the end user.  Other entry points
    /// will be used such as when the application is launched to open a specific file.
    /// </summary>
    /// <param name="args">Details about the launch request and process.</param>
    protected override void OnLaunched(LaunchActivatedEventArgs args)
    {
        Resources.Build(r => r.Merged(
            new AppResources()));

        base.OnLaunched(args);

        MainWindow.SetWindowIcon();
    }

    /*****  NEW TEMPLATE BOILERPLATE ****************************************************/
    protected override void RegisterYourPlatformSpecificServices(IHostBuilder hostBuilder, IServiceCollection services)
    {
        base.RegisterYourPlatformSpecificServices(hostBuilder, services);
    }
}

```
